evm/p2p/peer.py

```diff
@@ -403,7 +418,10 @@ def sync(self, head_hash, head_number, reorg_depth=0):
                 # TODO: Implement a retry mechanism and only disconnect after retrying a few
                 # times.
                 self.logger.warn("Timeout when syncing, disconnecting from %s", self)
-                yield from self.stop()
+                yield from self.stop_and_wait_until_finished()
+            except Exception:
+                self.logger.error(
```

Comment:
Is there a reason this doesn't re-raise?

Reply:
Because this method is called with an asyncio.ensure_future(), so that exception would not be caught anywhere. This will change soon, though, as I'll implement a global (cross-peer) queue of sync requests to avoid fetching the same headers from multiple peers, and then those sync() calls will run within a loop where we'll be able to catch any exceptions they raise